### ISSUE TITLE
table.isEmpty generates incorrect SQL for (left/right) join queries

### ISSUE THREAD

**Issue by 0xfaded** [slick/slick#1300]
```
for ((foo, bar) <- Foo joinLeft Bar on (_.id === _.id) if bar.isEmpty) yield foo.id
```

generates SQL equivalent to

```
select x2.one, x1.id, x2.id from (select id from foo) x1 left outer join (select 1 as one, id from bar) x2 on x1.id = x2.id where not (x2.one = 1);
```

I can't find it in the SQL spec, but both mysql and sqlite treat `not (x2.one = 1)` as `x2.one <> 1`. Therefore, regardless of whether the right side of the join `isDefined` (`x2.one` is `1`) or `isEmpty` (`x2.one` is `NULL`), `x2.one <> 1` is false.

I suspect compiling to `x2.one is (not) null` is the fix here.

**Comment by 0xfaded** [comment]
```
create table foo (id int not null primary key);
create table bar (id int not null primary key);
insert into foo values (1);
select x2.one, x1.id, x2.id from (select id from foo) x1 left outer join (select 1 as one, id from bar) x2 on x1.id = x2.id;
select x2.one, x1.id, x2.id from (select id from foo) x1 left outer join (select 1 as one, id from bar) x2 on x1.id = x2.id where not (x2.one = 1);
```

Test script

**Comment by szeiger** [comment]
Which Slick version? This was fixed recently (but it's possible that there's another case in a similar situation)

**Comment by 0xfaded** [comment]
2.11

Apologies, I've been using play-slick and according to the documentation

```
Add a library dependency on play-slick:

"com.typesafe.play" %% "play-slick" % "1.0.1"
The above dependency will also bring along the Slick library as a transitive dependency. This implies you don’t need to add an explicit dependency on Slick.
```

should have brought in 3.0. I'll work out whats going on. Thank you

**Comment by 0xfaded** [comment]
Ok I've worked it out

I was using `slick_2.11 %% slick-3.0.1`

So this is present in `3.0.1`

**Comment by szeiger** [comment]
Should be fixed in Slick 3.1.0-RC2 (play-slick 1.1.0-RC2).

**Comment by gwak** [comment]
Hi, 
Sorry to report that this is still not fixed for me (or I didn't understand how to do it properly in slick). 
I'm using the latest version of play-slick (1.1.0-RC2).

Here's my query:
```scala
TableAs
    .joinLeft(TableBs).on((a, b) ⇒ a.colA === b.colA && a.colB === b.colB)
    .filter(_._2.isEmpty)
    .map(r ⇒ (r._1.colA, r._1.colB))
```
Here's the generated SQL: 
```sql
select x2."col_a", x2."col_b"
from "table_a" x2
left outer join "table_b" x3
on (x2."col_a" = x3."col_a") and (x2."col_b" = x3."col_b")
where false
```

My workaround for now is using the following query instead:
 ```scala
TableAs
    .joinLeft(TableBs).on((a, b) ⇒ a.colA === b.colA && a.colB === b.colB)
    .filter(_._2.flatMap(_.colA).isEmpty)
    .map(r ⇒ (r._1.colA, r._1.colB))
```

Which generates the following SQL (it's not ideal, but it works as expected in PostgreSql):
```sql
select x2."col_a", x2."col_b"
from "table_a" x2
left outer join "table_b" x3
on (x2."col_a" = x3."col_a") and (x2."col_b" = x3."col_b")
where (case when (not false) then x3."col_a" else null end) is null
```
